server/api/permission/list/service.js

```diff
@@ -36,7 +36,17 @@ async function listPermissions(db, { q, limit, cursor, scopes }) {
                 $gt: ObjectId(cursor.id),
               },
             }
-          : {}
+          : {},
+        role
```

Comment:
Hey, I thought you didn't like this pattern so match. Feel free to use an external object and decorate it accordingly, e.g.

```javascript
const matchExpressions = [];

if (!scopes.includes(null)) {
  matchExpressions.push({
    scope: { $in: scopes.map((scope) => ObjectId(scope)) }
  });
}

if (q) {
  matchExpressions.push({
    name: {
      $regex: `^${escapeRegExp(q)}`,
      $options: 'i',
    }
  });
}

if (cursor) {
  matchExpressions.push({
    _id: { $gt: ObjectId(cursor.id) }
  });
}

if (role) {
  matchExpressions.push({
    _id: { $in: role.permissions.map((permission) => ObjectId(permission)) }
  });
}

if (isSystemPermission) {
  matchExpressions.push({
    isSystemPermission: { $eq: isSystemPermission },
  });
}
```

And then you need only do...

```javascript
const permissions = await PermissionModel.aggregate([
    {
      $match: { $and: matchExpressions }
    }
    // aggregation pipeline follows
]);
```